src/transform/src/fusion/join.rs

```diff
@@ -50,85 +50,152 @@ impl Join {
-                    *relation = MirRelationExpr::constant(vec![vec![]], RelationType::empty());
-                }
-                1 => {
-                    // if there are constraints, they probably should have
-                    // been pushed down by predicate pushdown, but .. let's
-                    // not re-write that code here.
-                    if equivalences.is_empty() {
-                        *relation = inputs.pop().unwrap();
+            *relation = join_builder.build();
+        }
+    }
+}
+
+/// Helper builder for fusing the inputs of nested joins into a single Join expression.
+struct JoinBuilder {
+    inputs: Vec<MirRelationExpr>,
+    equivalences: Vec<Vec<MirScalarExpr>>,
+    num_columns: usize,
+    /// Predicates that will be evaluated on top of the join, if any.
+    predicates: Vec<MirScalarExpr>,
+}
+
+impl JoinBuilder {
+    fn new(equivalences: &mut Vec<Vec<MirScalarExpr>>) -> Self {
+        Self {
+            inputs: Vec::new(),
+            equivalences: equivalences.drain(..).collect(),
+            num_columns: 0,
+            predicates: Vec::new(),
+        }
+    }
+
+    fn add_input(&mut self, input: MirRelationExpr) {
+        self.num_columns += input.arity();
+        self.inputs.push(input);
+    }
+
+    fn add_inputs<I>(
+        &mut self,
+        inputs: I,
+        mut equivalences: Vec<Vec<MirScalarExpr>>,
+        predicates: Option<Vec<MirScalarExpr>>,
+    ) where
+        I: IntoIterator<Item = MirRelationExpr>,
+    {
+        // Update and push all of the variables.
+        for mut equivalence in equivalences.drain(..) {
+            for expr in equivalence.iter_mut() {
+                expr.visit_mut(&mut |e| {
+                    if let MirScalarExpr::Column(c) = e {
+                        *c += self.num_columns;
                     }
-                }
-                _ => {}
+                });
             }
+            self.equivalences.push(equivalence);
+        }
+
+        if let Some(mut predicates) = predicates {
+            for mut expr in predicates.drain(..) {
+                expr.visit_mut(&mut |e| {
+                    if let MirScalarExpr::Column(c) = e {
+                        *c += self.num_columns;
+                    }
+                });
+                self.predicates.push(expr);
+            }
+        }
+
+        // Add all of the inputs.
+        for input in inputs {
+            self.add_input(input);
+        }
+    }
+
+    fn build(mut self) -> MirRelationExpr {
+        // Filter join identities out of the inputs.
+        // The join identity is a single 0-ary row constant expression.
+        self.inputs.retain(|input| {
+            if let MirRelationExpr::Constant {
+                rows: Ok(rows),
+                typ,
+            } = &input
+            {
+                !(rows.len() == 1 && typ.column_types.len() == 0 && rows[0].1 == 1)
+            } else {
+                true
+            }
+        });
```

Comment:
I think this can be moved to `add_input`. As in, we can just avoid pushing an input into `self.inputs` if it is a join identity.

Reply:
right! thanks!